library-model/library-model-generator-cli/src/main/java/com/uber/nullaway/libmodel/LibraryModelGenerator.java

```diff
@@ -0,0 +1,295 @@
+    private static final String JSPECIFY_NULLABLE_IMPORT = "org.jspecify.annotations.Nullable";
+
+    public AnnotationCollectionVisitor(Map<String, MethodAnnotationsRecord> methodRecords) {
+      this.methodRecords = methodRecords;
+    }
+
+    @Override
+    public void visit(PackageDeclaration pd, Void arg) {
+      this.packageName = pd.getNameAsString();
+      super.visit(pd, null);
+    }
+
+    @Override
+    public void visit(ImportDeclaration id, Void arg) {
+      if (id.getName().toString().contains(JSPECIFY_NULLABLE_IMPORT)) {
+        this.isJspecifyNullableImportPresent = true;
+      }
+      super.visit(id, null);
+    }
+
+    @Override
+    public void visit(ClassOrInterfaceDeclaration cid, Void arg) {
+      String classOrInterfaceName = packageName + "." + cid.getNameAsString();
+      @SuppressWarnings("all")
+      Map<Integer, String> nullableTypeBoundsMap = new HashMap<>();
+      List<TypeParameter> paramList = cid.getTypeParameters();
+      // Finding Nullable upper bounds for generic type parameters.
+      for (int i = 0; i < paramList.size(); i++) {
+        boolean hasNullableUpperBound = false;
+        NodeList<ClassOrInterfaceType> upperBoundList = paramList.get(i).getTypeBound();
+        for (ClassOrInterfaceType upperBound : upperBoundList) {
+          if (upperBound.isAnnotationPresent("Nullable")) {
+            hasNullableUpperBound = true;
+            break;
+          }
+        }
+        if (hasNullableUpperBound) {
+          nullableTypeBoundsMap.put(i, classOrInterfaceName);
+        }
+      }
+      super.visit(cid, null);
+    }
+
+    @Override
+    public void visit(MethodDeclaration md, Void arg) {
+      Optional<Node> parentClassNode = md.getParentNode();
+      String parentClassName = "";
+      if (parentClassNode.isPresent()) {
+        if (parentClassNode.get() instanceof ClassOrInterfaceDeclaration) {
+          parentClassName = ((ClassOrInterfaceDeclaration) parentClassNode.get()).getNameAsString();
+        } else if (parentClassNode.get() instanceof EnumDeclaration) {
+          parentClassName = ((EnumDeclaration) parentClassNode.get()).getNameAsString();
+        }
+      }
+      if (hasNullableReturn(md)) {
+        methodRecords.put(
+            packageName
+                + "."
+                + parentClassName
+                + ":"
+                + getMethodReturnTypeString(md)
+                + " "
+                + md.getSignature().toString(),
+            MethodAnnotationsRecord.create(ImmutableSet.of("Nullable"), ImmutableMap.of()));
+      }
+      super.visit(md, null);
+    }
+
+    /**
+     * Determines if a MethodDeclaration can return null.
+     *
+     * @param md The MethodDeclaration instance.
+     * @return {@code true} if the method can return null, {@code false} otherwise.
+     */
+    private boolean hasNullableReturn(MethodDeclaration md) {
+      if (md.getType() instanceof ArrayType) {
```

Comment:
Add a comment explaining this code and why we do something different for array types